packages/loader/container-loader/src/test/serializedStateManager.spec.ts

```diff
@@ -634,4 +432,448 @@ describe("serializedStateManager", () => {
+				pending,
+				logger.toTelemetryLogger(),
+				storageAdapter,
+				true,
+				eventEmitter,
+				() => isDirty,
+			);
+			const snapshotSequenceNumber = 11;
+			storageAdapter.uploadSummary(snapshotSequenceNumber);
+
+			await serializedStateManager.fetchSnapshot(undefined, false);
+			await serializedStateManager.waitForInitialRefresh;
+			const state = await serializedStateManager.getPendingLocalStateCore(
+				{ notifyImminentClosure: false },
+				"clientId",
+				new MockRuntime(),
+				resolvedUrl,
+			);
+			const parsed = JSON.parse(state) as IPendingContainerState;
+			assert.strictEqual(parsed.baseSnapshot.id, "fromPending");
+			assert.strictEqual(parsed.savedOps.length, 0, "should not be savedOps");
+		});
+
+		it(`refresh snapshot when snapshot sequence number is above processed ops, isDirty: ${isDirty}`, async () => {
+			const pending: IPendingContainerState = {
+				...pendingLocalState,
+				baseSnapshot: { ...snapshot, id: "fromPending" },
+			};
+			let saved = false;
+			const isDirtyF = () => (saved ? false : isDirty);
+			const storageAdapter = new MockStorageAdapter();
+			const serializedStateManager = new SerializedStateManager(
+				pending,
+				logger.toTelemetryLogger(),
+				storageAdapter,
+				true,
+				eventEmitter,
+				isDirtyF,
+			);
+			let seq = 1;
+			let lastProcessedOpSequenceNumber = 20;
+			while (seq <= lastProcessedOpSequenceNumber) {
+				serializedStateManager.addProcessedOp(generateSavedOp(seq++));
+			}
+
+			const snapshotSequenceNumber = 30;
+			storageAdapter.uploadSummary(snapshotSequenceNumber);
+
+			await serializedStateManager.fetchSnapshot(undefined, false);
+			// latest snapshot fetched but we're still behind the snapshot.
+			// next addProcessedOp calls will be responsible for refreshing
+			await serializedStateManager.waitForInitialRefresh;
+
+			lastProcessedOpSequenceNumber = 40;
+			while (seq <= lastProcessedOpSequenceNumber) {
+				serializedStateManager.addProcessedOp(generateSavedOp(seq++));
+			}
+			if (isDirty) {
+				saved = true;
+				eventEmitter.emit("saved");
+			}
+
+			const state = await serializedStateManager.getPendingLocalStateCore(
+				{ notifyImminentClosure: false },
+				"clientId",
+				new MockRuntime(),
+				resolvedUrl,
+			);
+			const parsed = JSON.parse(state) as IPendingContainerState;
+			assert.strictEqual(parsed.baseSnapshot.id, "fromStorage");
+			const attributes = getAttributesFromPendingState(parsed);
+			assert.strictEqual(
+				attributes.sequenceNumber,
+				snapshotSequenceNumber,
+				"wrong snapshot sequence number",
+			);
+			assert.strictEqual(
+				parsed.savedOps[0].sequenceNumber,
+				snapshotSequenceNumber + 1,
+				"wrong first saved op",
+			);
+			assert.strictEqual(
+				parsed.savedOps[parsed.savedOps.length - 1].sequenceNumber,
+				lastProcessedOpSequenceNumber,
+				"wrong last saved op",
+			);
+		});
+
+		it(`does not refresh snapshot when we haven't processed all saved ops, isDirty: ${isDirty}`, async () => {
```

Comment:
new tests to check edge case scenario when we meet all checks except processing the saved ops